src/dict.c

```diff
@@ -61,6 +61,7 @@
  * the number of elements and the buckets > dict_force_resize_ratio. */
 static int dict_can_resize = 1;
 static unsigned int dict_force_resize_ratio = 5;
+static dictExpandCheckerFunction dictExpandChecker = NULL;
```

Comment:
Actually, I also care about this question, in practice, I occurred very big hash objects, I think we should take care it, and that will waste some memory if we add a member function of dict. Initially I thought we can check only when dicts exceed a certain size, such as 128 or 256. Just as you said, there may be many small hash objects, so I am not sure, but it is small possibility.

Reply:
i didn't mean to add another pointer to each hash, i was thinking of passing some argument from the caller.
i was thinking of `tryResizeHashTables` in server.c (which triggers rehashing only for the db dict and expires dict).
but i now realize that this only controls shrinking, and that expanding happens implicitly.
still we may be able to find some 1 bit flag to disable the implicit behavior for these 2 dicts, and have it controlled from the caller.

but first let's finish the debate on whatever we want just these two fields to be affected by the new limit, or have them all affected (maybe with a size threshold).

Reply:
@redis/core-team we need your feedback. do you think such a mechanism needs to work on all hash tables? just the main db and expires dict? or just hash tables above a certain size?

note that there might be a use case where a lot of small hash tables decide to rehash together and cause eviction too. but on the other hand a more common case is that it happens with just the main dict, and when it does happen, we don't want to enforce this limit on small hash tables.. (if we do, maybe we also want to prevent them from being converted from ziplist to dict).

my feeling is that we need to try to solve just the common problem we're aware of, without causing any unexpected side effects which we might not be fully aware of.

Reply:
My opinion, it should just be on the main DB:
1. This should be an API on the type, not a single static variable so it's at least configurable. I also don't like the API, since I still think people copy code out of Redis, I think the API should be clean. The API should also pass in the dict and the new power, and we should do our own memory computation. 
2. Redis has a soft maxmemory, so it's possible we go through all this work to not-rehash, to still exceed the max memory and require evictions. It's weird that we now have this weird internal logic where we are trying to not go over it. Adding these types of assumptions is complexity we have to maintain, and I don't set the justification for most of the dictionaries. Just solving the known problem makes this as transparent as possible. 
3. Imagine having a volatile-lru memory policy, with a hash in the non-volatile part of memory. The user might assume volatile keys will get evicted as they grow the hash. This is true, but we won't rehash the dict, so the dict will slow down. The Redis hash has a load factor of 1 so it's fast, we could end up with a load factor of 10+, killing performance.